Follow-up to #478

scober opened the pull request:
I rushed to merge #478 because it was blocking some people. This PR addresses the outstanding issues from reviewers of that PR.

For the issue of moving the templated function to a cpp file I decided to put most of the logic in a helper function and put the helper function in a cpp file. I didn't want to pass a raw pointer without a suitable buffer class to wrap it in, so this seemed like the best I could do. If we do settle on a suitable buffer class, I am happy to move to that implementation.

ddn0 commented:
Working on fixing the conda CI issue, but that's unrelated to the C++ changes.

Ideally, the ephemeral properties or similar will address using non-trivially constructable properties. We probably want some interface that allows for one definition of how to serialize data that is used both for persisting to disk and for libgluon/libdist communication.

roshandathathri commented on lonestar/analytics/distributed/bfs/bfs_push.cpp:
```diff
@@ -71,16 +71,17 @@ static cll::opt<Exec> execution(
-  NODE_DATA_DIST_OLD,
+struct NodeDistCurrent {
+  using ArrowType = arrow::CTypeTraits<uint32_t>::ArrowType;
+  using ViewType  = galois::PODPropertyView<std::atomic<uint32_t>>;
```
Can we please have a PODAtomicProperty to simplify this?

ddn0 replied:
Do we really want it to be easier to use `std::atomic` all the time? `std::atomic_ref` is the more forward looking interface though not standard until C++20. https://en.cppreference.com/w/cpp/atomic/atomic_ref 

My preference is to leave this as it is. These types of changes are easy to add later and incrementally without breaking existing code if we really think this is the right way to go.

roshandathathri replied:
Ok, makes sense. This can be deferred.

I didn't know about atomic_ref. That's preferable. It seems like they provide compare-and-swap to any trivially-copyable type? How would that be achieved if it does not fit in an instructions's word?

To answer your question, I think we should make it easier for users to write concurrent code, which includes transparently handling atomics.

ddn0 replied:
> It seems like they provide compare-and-swap to any trivially-copyable type? How would that be achieved if it does not fit in an instructions's word?

Same as `std::atomic` apparently:

https://stackoverflow.com/questions/58346307/how-is-stdatomic-ref-implemented-for-non-atomic-types

This is all news to me, as I don't usually think of using `std::atomic` for non-word things.

roshandathathri replied:
Wow. It is handling locks for pointers transparently. Not sure how efficient it will be because the number of locks required in our case would be the number of nodes. But it might be worth trying out as it saves us from the effort of managing locks. Specifically, we (@l-hoang) used pull-style for MRBC because push-style required atomically updated two variables. This might also be useful bc-async (which is only implemented for shared-memory), which currently uses more coarse-grained locks than required in my opinion.

witchel replied:
Also news to me and also vaguely alarming in terms of storage overhead.  I guess the semantics are clean, but the whole "if you aren't naturally aligned you aren't atomic silently" caveat didn't fill me with joy.

roshandathathri reviewed (approved):
LGTM

Some minor nits

ddn0 commented on libgluon/include/galois/graphs/GluonSubstrate.h:
```diff
@@ -3233,21 +3233,22 @@ class GluonSubstrate : public galois::runtime::GlobalObject {
-  template <bool isVoidType,
-            typename std::enable_if<isVoidType>::type* = nullptr>
+  template <typename U                                            = EdgeTy,
+            typename std::enable_if_t<std::tuple_size_v<U> == 0>* = nullptr>
```
`EdgeTy` is defined as `using EdgeTy = typename GraphTy::EdgeType;`. So, this is inferred from GraphTy, transitively. Are you asking to change the above to `template <typename U = typename GraphTy::EdgeType, ....>`  and replace all references to `EdgeTy`?

Otherwise, do you mean defining `setMarshalEdge` without `template <typename U>` at all. It is possible if you don't use `std::enable_if`. `std::enable_if` relies on SFINAE so `template <typename U>` is required to defer instantiation to after class parsing.

I've updated the PR to use `if constexpr` instead. This is not my preferred way of doing template metaprogramming because `if constexpr` doesn't generalize well and leads to undisciplined metaprogramming.

roshandathathri replied:
I had missed that it was inferred transitively, so this is fine.

I prefer the constexpr way. I'm curious as to why it doesn't generalize well?

ddn0 replied:
- example of undisciplined: `if constexpr` sprinkled throughout instead of encouraging developers to factor template dispatch along function boundaries that hopefully align better along concepts being distpatched on (e.g., supports random access)
- example of not generalizing well: chains of `if constexpr () {} else if constexpr () {}` to do multiway switches rather than function overloading or some other kind of abstraction if possible.

I think `if constexpr` is reasonable for a single boolean and probably more readable in this instance (and I've cleaned up `Traits.h` to use it as well) but if I was instead optimizing for minimizing the amount of C++ knowledge required to understand and contribute to a codebase, I'd probably avoid mentioning `if constexpr` at all.

roshandathathri replied:
Thanks. I agree with the points.

I expect `std::enable_if` is more difficult to parse or understand for a novice than `if constexpr`. Maybe that's just me.

gurbinder533 commented:
@ddn0 **bc_mr.cpp** has gstl::Vector and MRBCTree data structure along with some PODs as the node data. NodeData is fairly complex. I am not sure what the ArrowType and ViewType should be for these. Do we need to refactor bc_mr and simplify node data to use arrowArrays?

ddn0 commented:
> bc_mr.cpp has gstl::Vector and MRBCTree data structure along with some PODs as the node data. NodeData is fairly complex. I am not sure what the ArrowType and ViewType should be for these. Do we need to refactor bc_mr and simplify node data to use arrowArrays?

One plan of attack would be:
1. add a "ByteArray<size_t>" arrow type to serve as the backing storage of MRBCTree
2. construct MRBCTree using a placement new
3. use MRBCTree is the view type

This raises some questions like how would a nested `std::vector` in MRBCTree be persisted or sent to another host for distribution? Ideally, we'd have some serialize and deserialize to a byte buffer interface and use that for disk and for message passing. We have the `gSerialize` stuff, though some might prefer a separate data description language/specification (a la protobuf, arrow, thrift etc.) that would allow other programs to read serialized buffers rather than executing the original application.

You get into similar questions when you want to persist an index: how do we want to serialize and deserialize non-trivial data structures (and probably support their evolution separate from the application that initially created them)? Nowish, is probably a good time to start thinking about what we want to do here in general. It could be the case that we decide to treat these problems separately, but it seems like having an agreed-upon method for serializing/deserializing data will help us all.

We have some initial work saving and restoring partitions, `gSerialize`, Gluon, the partitioner, so I think we should have enough experience to make a good design.

cc @witchel , @tylershunt

Commit:
Move bulk of AssignEntityTypeIDsFromProperties to cpp file

The whole function can't be moved because it is templated, but this gets
most of the logic into a cpp file.

An alternative would have been to make this a non-template function and
accept a C-style array instead of a template-defined vector type. Then
the entire function could have been moved to the cpp file.

A variation of this would be to create a class that wraps C-style arrays
and tracks the size and offers begin(), end(), etc.

Commit:
Drop EntityTypeManager::Reset()

It was unused and unnecessary. Any use of Reset() can be replaced with
=EntityTypeManager{}, so the presence of Reset() in the interface is
at best confusing and at worst misleading.

amberhassaan commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -101,146 +92,47 @@ class EntityTypeManager {
   /// The number of entries in entity_type_ids should be equal to topo_size and
   /// properties->num_rows().
-  template <template <typename> class Vector>
+  template <typename EntityTypeVector>
```
@roshandathathri said we should name this EntityTypeArray instead of Vector.

amberhassaan commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -101,146 +92,47 @@ class EntityTypeManager {
   ///
   /// The number of entries in entity_type_ids should be equal to topo_size and
```
properties->num_rows() is allowed to be 0 even when entity_type_ids is not.

amberhassaan reviewed (approved):
Looks good. Left minor comments.

arthurp commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -83,164 +83,57 @@ class EntityTypeManager {
+    static_assert(
+        std::is_same_v<typename EntityTypeArray::value_type, EntityTypeID>);
+    if (entity_type_ids->size() != topo_size) {
+      return KATANA_ERROR(
+          katana::ErrorCode::InvalidArgument,
+          "size of topology array ({}) doesn't match size of type array ({})",
+          topo_size, entity_type_ids->size());
```
Yay!! Thanks for removing these asserts! I'm glad people are starting to think this way!

arthurp commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -471,6 +355,30 @@ class EntityTypeManager {
+  struct TypeProperties {
+    using BoolPropertyColumn = PropertyColumn<arrow::BooleanArray>;
+    std::vector<BoolPropertyColumn> bool_properties;
+    using UInt8PropertyColumn = PropertyColumn<arrow::UInt8Array>;
```
I think these two type aliases are only used once. And they are simple and private. I would inline them.

scober replied:
Good call

arthurp reviewed (approved):
LGTM. Only one very small nit.

arthurp commented:
https://github.com/KatanaGraph/katana/pull/459 will conflict with this. Tell me once this is merged and I'll rebase it and do the merge and those PRs will be ready.

scober commented:
@arthurp merged

amberhassaan commented:
April 4th, 1984

arthurp commented:
> April 4th, 1984

WTF?

amberhassaan commented:
The number of PR reminded me of that. And then I thought, we'll hit PR#1984 in enterprise soon.
